src/packet/udp.rs

```diff
@@ -101,6 +50,8 @@ fn udp_header_ipv4_test() {
     packet[20 + 8 + 3] = 't' as u8;
 
     {
+        use pnet::packet::ipv4::Ipv4Packet;
+        let ip_header = Ipv4Packet::new(&packet[..]).unwrap();
         let mut udp_header = MutableUdpPacket::new(&mut packet[20..]).unwrap();
```

Comment:
So the error you're getting here:
```
src/packet/udp.rs:55:57: 55:63 error: cannot borrow `packet` as mutable because it is also borrowed as immutable
src/packet/udp.rs:55         let mut udp_header = MutableUdpPacket::new(&mut packet[20..]).unwrap();
                                                                             ^~~~~~
src/packet/udp.rs:54:42: 54:48 note: previous borrow of `packet` occurs here; the immutable borrow prevents subsequent moves or mutable borrows of `packet` until the borrow ends
src/packet/udp.rs:54         let ip_header = Ipv4Packet::new(&packet[..]).unwrap();
                                                              ^~~~~~
src/packet/udp.rs:70:6: 70:6 note: previous borrow ends here
src/packet/udp.rs:52     {
...
src/packet/udp.rs:70     }
                         ^
```
Is exactly as it says - you cannot have both mutable and immutable references to a packet at the same time. This is the reason the old `checksum` function took a source and destination address, rather than taking an `Ipv4Packet`.

There are various ways to work around this, I don't know which is best without some additional thought though.